src/browser/cdp/selectivity/js-selectivity.ts

```diff
@@ -0,0 +1,178 @@
+    private _scriptsSource: Record<CDPRuntimeScriptId, null | Promise<string | Error>> = {};
+    private _scriptsSourceMap: Record<CDPRuntimeScriptId, null | Promise<string | Error>> = {};
+
+    constructor(cdp: CDP, sessionId: CDPSessionId, sourceRoot = "") {
+        this._cdp = cdp;
+        this._sessionId = sessionId;
+        this._sourceRoot = sourceRoot;
+    }
+
+    private _processScript({ scriptId, url, sourceMapURL }: DebuggerEvents["scriptParsed"]): void {
+        if (!this._sessionId) {
+            return;
+        }
+
+        if (!url || !sourceMapURL) {
+            this._scriptsSource[scriptId] ||= null;
+            this._scriptsSourceMap[scriptId] ||= null;
+            return;
+        }
+
+        this._scriptsSource[scriptId] ||= this._cdp.debugger
+            .getScriptSource(this._sessionId, scriptId)
+            .then(res => res.scriptSource)
+            .catch((err: Error) => err);
+
+        this._scriptsSourceMap[scriptId] ||= fetchTextWithBrowserFallback(
+            urlResolve(url, sourceMapURL),
+            this._cdp.runtime,
+            this._sessionId,
+        );
+    }
+
+    async start(): Promise<void> {
+        const debuggerOnPaused = (this._debuggerOnPausedFn = async (): Promise<void> => {
+            return this._cdp.debugger.resume(this._sessionId).catch(() => {});
+        });
+
+        const debuggerOnScriptParsedFn = (this._debuggerOnScriptParsedFn = this._processScript.bind(this));
+        const sessionId = this._sessionId;
+
+        this._cdp.debugger.on("paused", debuggerOnPaused);
+        this._cdp.debugger.on("scriptParsed", debuggerOnScriptParsedFn);
+
+        await Promise.all([
+            this._cdp.target.setAutoAttach(sessionId, { autoAttach: true, waitForDebuggerOnStart: false }),
+            this._cdp.debugger.enable(sessionId),
+            this._cdp.profiler.enable(sessionId).then(() =>
+                this._cdp.profiler.startPreciseCoverage(sessionId, {
+                    callCount: false,
+                    detailed: false,
+                    allowTriggeredUpdates: false,
+                }),
+            ),
+        ]);
+    }
+
+    /** @param drop only performs cleanup without providing actual deps. Should be "true" if test is failed */
+    async stop(drop?: boolean): Promise<string[]> {
+        if (drop) {
+            this._debuggerOnPausedFn && this._cdp.debugger.off("paused", this._debuggerOnPausedFn);
+            this._debuggerOnScriptParsedFn && this._cdp.debugger.off("scriptParsed", this._debuggerOnScriptParsedFn);
+
+            return [];
+        }
+
+        const coverage = await this._cdp.profiler.takePreciseCoverage(this._sessionId);
+
+        const scriptsWithUrl = coverage.result.filter(script => script.url);
+
+        // If we haven't got "scriptParsed" event for the script, pull up source code + source map manually
+        scriptsWithUrl.forEach(({ scriptId, url }) => {
```

Comment:
Unlike "scriptParsed" event, "takePreciseCoverage" can't be lost.
So here we ensure we didn't miss any "scriptParsed" event